### DISCUSSION TITLE
eCAL tools and ROS2 (with rmw_ecal and protobuf)

### DISCUSSION THREAD

**Discussion by sebastian-freitag** [eclipse-ecal/ecal#561] [Q&A]
Hi,

I am very curious about eCAL and played around a bit with https://github.com/continental/rmw_ecal and with ROS2. I got something running with rmw_ecal_proto_cpp and I could run ecal_mon_gui and see my topics. However, when I tried to inspect the topic and see if the protobuf reflection works, I get "ERROR [587]: CDynamicSubscriber: Could not get description for topic rt/topic" (when running simple talker nodes from the ROS2 samples). Can someone help me with troubleshooting? Is there some documentation about how the introspection works? Does ecal_mon_gui need the .proto messages from the ROS2 talker to make sense or is it completely relying on protobufs meta for introspection? And if yes, why does it work for the mma sample but not for my ROS2 node?

Cheers
Sebastian

**Comment by Kerstin-Keller** [comment]
In general, the monitor is relying completely on protobuf metadata, referred to as a `descriptor`. All compiled protobuf publishers / subscribers (as in eCAL samples https://github.com/continental/ecal/blob/master/samples/cpp/person/person_snd/src/person_snd.cpp ) register that information with eCAL, and thus the monitor can correctly display incoming messages without compiling the datatypes.

As far as I know, the rmw_ecal_proto_cpp layer should also register the respective infos with eCAL. 
@brakmic-aleksandar can you answer @sebastian-freitag-pylot question regarding `rmw_ecal_proto_cpp`?
However, please note that we don't consider the rmw_ecal as 100% stable. I wouldn't (yet) recommend it for production scenarios.

If you have the freedom, we would rather suggest a pure eCAL usage ;)

**Reply by sebastian-freitag** [reply]
Thanks for that first answer. So for the problem I see, something went wrong with the registration...

It's not entirely clear to me how the registration of the `descriptor` works. ecal is brokerless, so where is the registry located and roughly how does the registration mechanism work?

If there is already some technical documentation available, please point me to it. The documentation I found does not go into detail much about issues like this one. Also @brakmic-aleksandar , I would probably also be very happy to hear your comment about this. :+1:

**Reply by brakmic-aleksandar** [reply]
Yeah, rmw_ecal_proto_cpp should send metadata to monitoring layer automatically, maybe the fix you've made to avoid "File already exists in database" is causing a file descriptor get messed up (somehow) i'll try it out tomorrow to see whether it's causing the issue,

**Reply by Kerstin-Keller** [reply]
Yes, we need to improve technical documentation on that matter.
There is a so called "registration" layer, where every participant periodically broadcasts information over UDP about what topics it consumes /  produces. this information includes the meta information about the publishers, including the protobuf descriptors.
